runtime/lua/vim/filetype/options.lua

```diff
@@ -0,0 +1,122 @@
+    string.format('ftplugin/%s.lua', filetype),
+    string.format('ftplugin/%s_*.lua', filetype),
+    string.format('ftplugin/%s/*.lua', filetype),
+  }, true, {}) --[[@as string[] ]]
+end
+
+-- Keep track of ftplugin files
+local ftplugin_cache = {} ---@type table<string,table<string,integer>>
+
+-- Keep track of total number of FileType autocmds
+local ft_autocmd_num ---@type integer?
+
+-- Keep track of filetype options
+local ft_option_cache = {} ---@type table<string,table<string,any>>
+
+--- @param path string
+--- @return integer
+local function hash(path)
+  local mtime0 = vim.loop.fs_stat(path).mtime
+  return mtime0.sec * 1000000000 + mtime0.nsec
+end
+
+--- Only update the cache on changes to the number of FileType autocmds
+--- and changes to any ftplugin/ file. This isn't guaranteed to catch everything
+--- but should be good enough.
+--- @param filetype string
+local function update_ft_option_cache(filetype)
+  local new_ftautos = #api.nvim_get_autocmds({ event = 'FileType' })
+  if new_ftautos ~= ft_autocmd_num then
+    -- invalidate
+    ft_option_cache[filetype] = nil
+    ft_autocmd_num = new_ftautos
+  end
+
+  local files = get_ftplugin_runtime(filetype)
+
+  ftplugin_cache[filetype] = ftplugin_cache[filetype] or {}
+
+  if #files ~= #vim.tbl_keys(ftplugin_cache[filetype]) then
+    -- invalidate
+    ft_option_cache[filetype] = nil
+    ftplugin_cache[filetype] = {}
+  end
+
+  for _, f in ipairs(files) do
+    -- VIMRUNTIME should be static so shouldn't need to worry about it changing
+    if not vim.startswith(f, vim.env.VIMRUNTIME) then
+      local mtime = hash(f)
+      if ftplugin_cache[filetype][f] ~= mtime then
+        -- invalidate
+        ft_option_cache[filetype] = nil
+        ftplugin_cache[filetype][f] = mtime
+      end
+    end
+  end
+end
+
+--- @private
+--- @param filetype string Filetype
+--- @param option string Option name
+--- @return string|integer|boolean
+local function get_option(filetype, option)
+  -- Some filetypes might not set 'commentstring' so we may end up
+  -- returning 'commentstring' for another filetype if the buffer is re-used.
+  local buf = api.nvim_create_buf(false, true)
+
+  -- Prevent syntax/ from being loaded
+  vim.b[buf].ts_highlight = true
+
+  -- Change the filetype without triggering any autocmds
+  noautocmd(function()
+    vim.bo[buf].filetype = filetype
+  end)
+
+  -- Trigger the FileType autocmd in ftplugin.vim
+  api.nvim_buf_call(buf, function()
+    -- TODO(lewis6991): should we just execute filetypeplugin or all of FileType?
+    -- Cache is only coherent for ftplugin but users might want to configure filetypes in their
+    -- config.via basic FileType autocmds.
+    -- api.nvim_exec_autocmds('FileType', { group = 'filetypeplugin', buffer = buf })
+    api.nvim_exec_autocmds('FileType', { buffer = buf })
+  end)
+
+  local ret = vim.bo[buf][option] ---@type string|integer|boolean
+
+  api.nvim_buf_delete(buf, {})
```

Comment:
This seems to cause a problem for me in some situations. I get an error on `E565: Not allowed to change text or change window`. I use vim-commentary and have an expr keymap that overrides the default one that just calculates the commentstring before returning `<Plug>Commentary`. It could maybe have something to do with me calling the function at a wrong time, but it should probably be documented if it's not allowed to get an option in some cases(?)

Reply:
comment.lua

```lua
local api = vim.api

local function noautocmd(f)
	local ei = vim.o.eventignore
	vim.o.eventignore = "all"
	f()
	vim.o.eventignore = ei
end

local function get_option(filetype, option)
	-- Some filetypes might not set 'commentstring' so we may end up
	-- returning 'commentstring' for another filetype if the buffer is re-used.
	local buf = api.nvim_create_buf(false, true)

	-- Prevent syntax/ from being loaded
	vim.b[buf].ts_highlight = true

	-- Change the filetype without triggering any autocmds
	noautocmd(function()
		vim.bo[buf].filetype = filetype
	end)

	-- Trigger the FileType autocmd in ftplugin.vim
	api.nvim_buf_call(buf, function()
		-- TODO(lewis6991): should we just execute filetypeplugin or all of FileType?
		-- Cache is only coherent for ftplugin but users might want to configure filetypes in their
		-- config.via basic FileType autocmds.
		-- api.nvim_exec_autocmds('FileType', { group = 'filetypeplugin', buffer = buf })
		api.nvim_exec_autocmds("FileType", { buffer = buf })
	end)

	local ret = vim.bo[buf][option] ---@type string|integer|boolean

	api.nvim_buf_delete(buf, {})

	return ret
end

local function calculate_commentstring()
	return get_option("lua", "commentstring")
end

vim.keymap.set("n", "gcc", function()
	vim.bo.commentstring = calculate_commentstring()
	return "<Plug>CommentaryLine"
end, { expr = true })
```

`nvim -u comment.lua test.lua`.

Insert some text and try `gcc`

Reply:
Works for me. Are you sure you don't have any wacky FileType autocmds? Can you reproduce this in `nvim --clean`?

Reply:
Nevermind, can repro.

Reply:
How about if you uncomment this:

```lua
    -- api.nvim_exec_autocmds('FileType', { group = 'filetypeplugin', buffer = buf })
```

and remove the line below?

Reply:
Also wrap the buf deletion in schedule:

```lua
  -- Need to schedule the buf deletion in case we are in an expr mapping
  vim.schedule(function()
    api.nvim_buf_delete(buf, {})
  end)
```

Reply:
Wrapping it in vim.schedule did the trick for me. Both of the exec_autocmds seem to work then

Reply:
Also you can do this without an `expr` mapping.

```lua
vim.keymap.set("n", "gcc", function()
  local lang = get_lang() or vim.bo.filetype
  vim.bo.commentstring = get_ft_option(lang, 'commentstring')
  vim.cmd.normal(vim.api.nvim_replace_termcodes("<Plug>CommentaryLine", true, false, true))
end)
```

Reply:
> Wrapping it in vim.schedule did the trick for me. Both of the exec_autocmds seem to work then

The problem is, we can't predict what will be in a `FileType` autocmd. For example, I had `vim.cmd.normal'gx'` which triggered a similar warning.

 So we will need to document this shouldn't be used in expr keymaps (though it might work).

Reply:
> > Wrapping it in vim.schedule did the trick for me. Both of the exec_autocmds seem to work then
> 
> The problem is, we can't predict what will be in a `FileType` autocmd. For example, I had `vim.cmd.normal'gx'` which triggered a similar warning.
> 
> So we will need to document this shouldn't be used in expr keymaps (though it might work).

Ah okay, will switch to the replace_termcodes variant👍 thanks!

Reply:
@justinmk (and @bfredl) do you think it would be reasonable to override `textlock` and `ex_normal_lock` within `get_option()` because "we know what we're doing" (™) .

`eval_map_expr()` has this:

```c
  // Forbid changing text or using ":normal" to avoid most of the bad side
  // effects.  Also restore the cursor position.
  textlock++;
  ex_normal_lock++;
```

Which makes sense in the general case, but here could it be ok it bypass this? Especially since the mapping isn't really running in the scratch buffer.

Otherwise, we can just document the restriction.

Reply:
> reasonable to override textlock and ex_normal_lock within get_option() because "we know what we're doing" (™) .

Yeah that's a good idea. We should do that regardless of where we end up with this PR and https://github.com/neovim/neovim/pull/22681